Skill opportunity job

TheRealJessicaLi opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #[fill_in_number_here].
2. This PR does the following:
This PR adds Beam jobs for generating and deleting SkillOpportunityModels as well as relevant backend tests.

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

1. Run command `python -m scripts.run_backend_tests --test_target=core.jobs.batch_jobs.opportunity_management_jobs_test` to make sure the added unit tests pass.

Screenshot of the job on local server
![Screen Shot 2021-12-30 at 2 08 09 PM](https://user-images.githubusercontent.com/44102730/147781229-0c5a09cf-d1de-4a74-ba5c-0904ae302a2b.png)


## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

TheRealJessicaLi committed:
added job and basic tests

TheRealJessicaLi committed:
formatted

TheRealJessicaLi committed:
more lint fixes

TheRealJessicaLi committed:
more lint

TheRealJessicaLi committed:
lint

TheRealJessicaLi committed:
lint

TheRealJessicaLi committed:
lint

TheRealJessicaLi committed:
fixed some errors

TheRealJessicaLi committed:
attempting to fix lint

TheRealJessicaLi committed:
partially fixed test errors

TheRealJessicaLi committed:
lint

TheRealJessicaLi committed:
more lint:

TheRealJessicaLi committed:
more lint

TheRealJessicaLi committed:
small fixes

TheRealJessicaLi committed:
added tests and fixed pipeline

TheRealJessicaLi committed:
linted

TheRealJessicaLi committed:
lint

TheRealJessicaLi committed:
more lint

TheRealJessicaLi committed:
linted again

TheRealJessicaLi committed:
suppressed type warnings

TheRealJessicaLi committed:
more lint

TheRealJessicaLi committed:
lint

TheRealJessicaLi committed:
lint :)

TheRealJessicaLi committed:
lint fixes

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -35,25 +37,190 @@
+            | 'Delete all models' >> ndb_io.DeleteModels()
+        )
+
+        return (
+            skill_opportunity_model
+            | 'Count all new models' >> beam.combiners.Count.Globally()
+            | 'Only create result for new models when > 0' >> (
+                beam.Filter(lambda n: n > 0))
+            | 'Create result for new models' >> beam.Map(
+                lambda n: job_run_result.JobRunResult(stdout='SUCCESS %s' % n))
+        )
```
Please use `job_result_transforms.CountObjectsToJobRunResult()` here, look at the latest `DeleteExplorationOpportunitySummariesJob` to see how it is used.

TheRealJessicaLi replied:
Done, thanks!

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -35,25 +37,190 @@
+class GenerateSkillOpportunityModelJob(base_jobs.JobBase):
+    """Job for regenerating SkillOpportunityModel.
+
+    NOTE: The DeleteSkillOpportunityModelJob must be run before this
+    job.
+
+    """
```
```suggestion
    NOTE: The DeleteSkillOpportunityModelJob must be run before this
    job.
    """
```

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -35,25 +37,190 @@
+                question_count=skill_opportunity.question_count
+            )
+            skill_opportunity_model.update_timestamps()
+            return {
+                'status': 'SUCCESS',
+                'job_result': job_run_result.JobRunResult(stdout='SUCCESS'),
+                'model': skill_opportunity_model
+            }
+        except Exception as e:
+            return {
+                'status': 'FAILURE',
+                'job_result': job_run_result.JobRunResult(
+                    stderr='FAILURE: %s' % e),
+                'model': None
+            }
```
Return `Result` object here instead, look into `GenerateExplorationOpportunitySummariesJob` to see how is it used.

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -35,25 +37,190 @@
+
+    def count_unique_question_ids(
+        self,
+        question_skill_link_models:
+            list[list[question_models.QuestionSkillLinkModel]]
+        ) -> int:
```
```suggestion
        question_skill_link_models: List[List[
            question_models.QuestionSkillLinkModel
        ]]
    ) -> int:
```

vojtechjelinek replied:
Also why is it list of list and not just list?

TheRealJessicaLi replied:
Reformatted. This is a 2D list because the CoGroupByKey() is operating on 2 dictionaries that map keys to values that are lists (so the resultant list is a list of list).

vojtechjelinek replied:
Could you reorganize it around line 193 in a way that it is just a list?

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -35,25 +37,190 @@
+        return len(set(question_ids))
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        all_question_skill_link_models = (
+            self.pipeline
+            | 'Get all non-deleted QuestionSkillLinkModels' >> (
+                ndb_io.GetModels(
+                    question_models.QuestionSkillLinkModel
+                    .get_all(include_deleted=False))
+                )
+            | 'Group QuestionSkillLinkModels by skill id' >>
+                beam.GroupBy(lambda n: n.skill_id)
+        )
```
```suggestion
        question_skill_link_models = (
            self.pipeline
            | 'Get all non-deleted QuestionSkillLinkModels' >> (
                ndb_io.GetModels(
                    question_models.QuestionSkillLinkModel
                    .get_all(include_deleted=False))
                )
            | 'Group QuestionSkillLinkModels by skill id' >>
                beam.GroupBy(lambda n: n.skill_id)
        )
```

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -35,25 +37,190 @@
+                beam.GroupBy(lambda n: n.skill_id)
+        )
+
+        all_skill_models = (
```
```suggestion
        skills = (
```

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -35,25 +37,190 @@
+        opportunities_results = (
+            skills_with_question_counts
+            | beam.Map(
+                lambda n:
+                self._create_skill_opportunity_model(
+                    opportunity_domain.SkillOpportunity( # type: ignore[no-untyped-call]
+                        skill_id=n[1]['skills'][0][0].id,
+                        skill_description=n[1]['skills'][0][0].description,
+                        question_count=self.count_unique_question_ids(
+                            n[1]['question_skill_links']
+                        ))))
```
This is quite complicated I suggest moving this to a function.

TheRealJessicaLi replied:
Just to clarify, which part of this are you suggesting we move?

vojtechjelinek replied:
I think that the whole lambda should be transformed into a function.

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -35,25 +37,190 @@
+            | 'Group Skills by skill id' >> beam.GroupBy(lambda m: m.id)
+        )
+
+        skills_with_question_counts = (({
+            'skills': all_skill_models,
+            'question_skill_links': all_question_skill_link_models
+        })
+            | 'Merge by skill id' >> beam.CoGroupByKey()
+        )
```
```suggestion
        skills_with_question_counts = (
            {
                'skill': all_skill_models,
                'question_skill_links': all_question_skill_link_models
            }
            | 'Merge by skill id' >> beam.CoGroupByKey()
        )
```

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs_test.py:
```diff
@@ -32,20 +32,313 @@
+    def setUp(self) -> None:
+        super().setUp()
+
+        question_1_model = self.create_model(
+            question_models.QuestionModel,
+            id=self.QUESTION_1_ID,
+            question_state_data={},
+            language_code=constants.DEFAULT_LANGUAGE_CODE,
+            version=0,
+            linked_skill_ids=[self.SKILL_1_ID, self.SKILL_2_ID],
+            inapplicable_skill_misconception_ids=[],
+            question_state_data_schema_version=feconf
+                .CURRENT_STATE_SCHEMA_VERSION
+        )
+
+        question_2_model = self.create_model(
+            question_models.QuestionModel,
+            id=self.QUESTION_2_ID,
+            question_state_data={},
+            language_code=constants.DEFAULT_LANGUAGE_CODE,
+            version=0,
+            linked_skill_ids=[self.SKILL_1_ID, self.SKILL_2_ID],
+            inapplicable_skill_misconception_ids=[],
+            question_state_data_schema_version=feconf
+                .CURRENT_STATE_SCHEMA_VERSION
+        )
+        question_1_model.update_timestamps()
+        question_2_model.update_timestamps()
```
Do we even need these? I don't see them used anywhere?

TheRealJessicaLi replied:
You're right, that was an oversight. Thanks!

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs_test.py:
```diff
@@ -32,20 +32,313 @@
+        question_1_model.update_timestamps()
+        question_2_model.update_timestamps()
+
+        question_1_skilllinkmodel = self.create_model(
+            question_models.QuestionSkillLinkModel,
+            question_id=self.QUESTION_1_ID,
+            skill_id=self.SKILL_1_ID,
+            skill_difficulty=1
+        )
```
```suggestion
        question_skill_link_model_1 = self.create_model(
            question_models.QuestionSkillLinkModel,
            question_id=self.QUESTION_1_ID,
            skill_id=self.SKILL_1_ID,
            skill_difficulty=1
        )
```

ditto for the second one

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs_test.py:
```diff
@@ -32,20 +32,313 @@
+        skill_1_model.update_timestamps()
+        skill_2_model.update_timestamps()
+
+        datastore_services.put_multi(
+            [skill_1_model, skill_2_model,
+             question_1_model, question_2_model,
+             question_1_skilllinkmodel,
+             question_2_skilllinkmodel]
+        )
```
```suggestion
        datastore_services.put_multi([
            skill_1_model,
            skill_2_model,
            question_1_model, 
            question_2_model,
            question_1_skilllinkmodel,
            question_2_skilllinkmodel
        ])
```

TheRealJessicaLi replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

TheRealJessicaLi committed:
merged with develop

TheRealJessicaLi committed:
use Result instead of JobRunResult in static helper methods

TheRealJessicaLi committed:
lint

TheRealJessicaLi committed:
fixed as per comments

vojtechjelinek commented:
@TheRealJessicaLi Please do not resolve comments yourself. Let the reviewer resolve them if they think that the comment was resolved.

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,172 @@
+            result.Result[opportunity_models.SkillOpportunityModel, Exception].
+            Metadata about the operation. Returns SkillOpportunityModel, or
+            Exception if an exception occurs.
+        """
```
```suggestion
        """Transform skill opportunity objects into skill opportunity models.

        Args:
            skill_opportunity: SkillOpportunity. The domain object.

        Returns:
            Result[opportunity_models.SkillOpportunityModel, Exception].
            Result object that contains SkillOpportunityModel when the operation
            is successful and Exception when an exception occurs.
        """
```

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,172 @@
+        question_skill_link_models: List[List[
+                question_models.QuestionSkillLinkModel
+            ]]
+        ) -> int:
```
```suggestion
    def count_unique_question_ids(
        self,
        question_skill_link_models: List[List[
            question_models.QuestionSkillLinkModel
        ]]
    ) -> int:
```

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,172 @@
+
+        Returns:
+            int. The number of unique question ids.
+        """
```
```suggestion
        """Counts the number of unique question IDs.

        Args:
            question_skill_link_models: list(list(questionSkillLinkModel)).
                2D array of QuestionSkillLinkModels.

        Returns:
            int. The number of unique question IDs.
        """
```

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,172 @@
+            int. The number of unique question ids.
+        """
+
+        question_ids = [link.question_id
+                        for link_list in question_skill_link_models
+                        for link in link_list]
+        return len(set(question_ids))
```
```suggestion
        question_ids = {
            link.question_id
            for link_list in question_skill_link_models 
            for link in link_list
        }
        return len(question_ids)
```

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,172 @@
+                        for link in link_list]
+        return len(set(question_ids))
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
```
Missing docstring.

TheRealJessicaLi replied:
Added, thanks!

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,172 @@
+            self.pipeline
+            | 'Get all non-deleted QuestionSkillLinkModels' >> (
+                ndb_io.GetModels(
+                    question_models.QuestionSkillLinkModel
+                    .get_all(include_deleted=False))
```
```suggestion
                    question_models.QuestionSkillLinkModel.get_all(
                        include_deleted=False))
```

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,172 @@
+
+        unused_put_result = (
+            opportunities_results
+            | 'Filter the results with SUCCESS status' >> beam.Filter(
```
Maybe update the explanation.

TheRealJessicaLi replied:
Done, thanks!

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,172 @@
+                lambda result: result.is_ok())
+            | 'Fetch the models to be put' >> beam.FlatMap(
+                lambda result: result.unwrap())
+            | 'Add ID as a key' >> beam.WithKeys(lambda model: model.id)  # pylint: disable=no-value-for-parameter
+            | 'Allow only one item per key' >> (
+                beam.combiners.Sample.FixedSizePerKey(1))
+            | 'Remove the IDs' >> beam.Values()  # pylint: disable=no-value-for-parameter
```
Why is this needed?

TheRealJessicaLi replied:
Whoops my bad, I missed this when copying over from the other job in the file. I don't think this is necessary. Thanks for pointing it out!

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,172 @@
+        return (
+            opportunities_results
+            | 'Fetch the job results' >> beam.Map(
+                lambda result: result['job_result'])
```
Wait, this is wrong right? `result` is `Result` object.

TheRealJessicaLi replied:
Good catch, thanks!

vojtechjelinek reviewed (changes requested):
Thanks! Took another pass.

TheRealJessicaLi committed:
Reformat doc string as per suggestion

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

TheRealJessicaLi committed:
Update spacing as per suggestion

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

TheRealJessicaLi committed:
Use round brackets instead of square brackets as per suggestion

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

TheRealJessicaLi committed:
Reformat question_ids as per suggestion

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

TheRealJessicaLi committed:
added doc string for run method

TheRealJessicaLi committed:
fixed formatting as per suggestion

TheRealJessicaLi committed:
Merge branch 'develop' into skill-opportunity-model-job

TheRealJessicaLi committed:
remove unecessary steps

TheRealJessicaLi committed:
fixed transformation from Results to JobRunResults

TheRealJessicaLi committed:
abstracted lambda for creating SkillOpportunityModel into a function

TheRealJessicaLi committed:
resolved conflicts

TheRealJessicaLi committed:
lint

TheRealJessicaLi committed:
more lint

TheRealJessicaLi committed:
fixing map for cogroupbykey so 'skill' maps to a Skill (not list)

TheRealJessicaLi committed:
fixed tests to expect correct string, and fixed job to use map instead of flatmap

TheRealJessicaLi committed:
removed lambda

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,179 @@
 
 (
     exp_models, opportunity_models, story_models,
-    topic_models
+    topic_models, skill_models, question_models
```
Please sort these alphabetically.

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,179 @@
+    """
+
+    @staticmethod
+    def _count_unique_question_ids(
+        question_skill_link_models: list[
+                question_models.QuestionSkillLinkModel
+            ]
+        ) -> int:
```
```suggestion
    def _count_unique_question_ids(
        question_skill_link_models: List[
            question_models.QuestionSkillLinkModel
        ]
    ) -> int:
```

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,179 @@
+            int. The number of unique question ids.
+        """
+
+        question_ids = [link.question_id for link in question_skill_link_models]
+        return len(set(question_ids))
```
```suggestion
        return len({link.question_id for link in question_skill_link_models})
```

vojtechjelinek replied:
Not done.

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,179 @@
+
+        Args:
+            question_skill_link_models: list(QuestionSkillLinkModel).
+                Array of QuestionSkillLinkModels.
```
```suggestion
                List of QuestionSkillLinkModels.
```

TheRealJessicaLi replied:
Whoops, my bad. Done.

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,179 @@
+    ) -> result.Result[
+        opportunity_models.SkillOpportunityModel, Exception
+    ]:
+        """Transform skill opportunity objects into skill opportunity models.
```
```suggestion
        """Transforms skill opportunity objects into skill opportunity models.
```

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,179 @@
+                    question_models.QuestionSkillLinkModel.get_all(
+                        include_deleted=False))
+                )
+            | 'Group QuestionSkillLinkModels by skill id' >>
```
```suggestion
            | 'Group QuestionSkillLinkModels by skill ID' >>
```

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,179 @@
+            | 'Get all non-deleted SkillModels' >> (
+                ndb_io.GetModels(
+                    skill_models.SkillModel.get_all(include_deleted=False)))
+            | 'Get skill from model' >> beam.Map(
+                skill_fetchers.get_skill_from_model)
+            | 'Group Skills by skill id' >> beam.GroupBy(lambda m: m.id)
```
```suggestion
            | 'Get skill object from model' >> beam.Map(
                skill_fetchers.get_skill_from_model)
            | 'Group skill objects by skill ID' >> beam.GroupBy(lambda m: m.id)
```

TheRealJessicaLi replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,179 @@
+                'skill': skills,
+                'question_skill_links': question_skill_link_models
+            }
+            | 'Merge by skill id' >> beam.CoGroupByKey()
+            | beam.Map(
+                lambda n: {
+                    'skill': n[1]['skill'][0][0],
+                    'question_skill_links': list(itertools.chain.from_iterable(
+                        n[1]['question_skill_links']))
+                }
+            )
```
```suggestion
            | 'Merge by skill ID' >> beam.CoGroupByKey()
            | 'Remove skill IDs' >> beam.Values()
            | beam.Map(
                lambda object: {
                    'skill': object['skill'][0][0],
                    'question_skill_links': list(itertools.chain.from_iterable(
                        object['question_skill_links']))
                }
            )
```

TheRealJessicaLi replied:
Is this necessary? The tests pass without beam.Values(), and I actually get a lint error when I add that line: "No value for argument 'pcoll' in function call"

vojtechjelinek replied:
I think it improves the readabiliy, you can add an ignore and comment explaining it 
![image](https://user-images.githubusercontent.com/10142938/146930124-68c1eef6-322b-4917-aea0-6da267559334.png)

TheRealJessicaLi replied:
Ok, done! Thanks

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,179 @@
+            | beam.Map(
+                lambda n: {
+                    'skill': n[1]['skill'][0][0],
+                    'question_skill_links': list(itertools.chain.from_iterable(
+                        n[1]['question_skill_links']))
```
Why we need `itertools.chain.from_iterable`?

TheRealJessicaLi replied:
I'm using this to flatten out the 2d list into a 1d list

vojtechjelinek replied:
Okay, can you add a comment above it?

TheRealJessicaLi replied:
Done!

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,179 @@
+        return len(set(question_ids))
+
+    @staticmethod
+    def _create_skill_opportunity_model(
+        skills_and_links: dict
+    ) -> result.Result[
```
Modify this to accept the skills and links as separate args and then down in the `run` method unpack the dictionary using lambda (`lambda object: _create_skill_opportunity_model(object['skill'], object['question_skill_links'])`), this should improve the readability of this function.

TheRealJessicaLi replied:
Good point, thanks! Done

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

TheRealJessicaLi committed:
Update spacing for _count_unique_question_ids signature

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

TheRealJessicaLi committed:
Add 'Remove skill IDs' in run()

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

TheRealJessicaLi committed:
fixed as per comments

TheRealJessicaLi committed:
fixed whitespace

TheRealJessicaLi committed:
Merge branch 'develop' into skill-opportunity-model-job

TheRealJessicaLi committed:
pulled and resolved merge conflicts

TheRealJessicaLi committed:
removed beam.Values()

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,187 @@
+            skill_opportunity = opportunity_domain.SkillOpportunity( # type: ignore[no-untyped-call]
+                skill_id=skill.id,
+                skill_description=skill.description,
+                question_count=GenerateSkillOpportunityModelJob
+                    ._count_unique_question_ids(question_skill_links)
```
Would this be possible?

```suggestion
                question_count=(
                    GenerateSkillOpportunityModelJob._count_unique_question_ids(
                        question_skill_links))
```

TheRealJessicaLi replied:
Done!

vojtechjelinek reviewed (changes requested):
Thanks! Few remaining comments.

TheRealJessicaLi committed:
fixed as per comments

TheRealJessicaLi committed:
linted

vojtechjelinek commented:
@TheRealJessicaLi Hey, can you run the local devserver and run the job there and add a screenshot of a successful run in the proof part. You can generate some skills using the admin page.

TheRealJessicaLi committed:
wrapped PCollection with list for indexing

TheRealJessicaLi committed:
Merge branch 'develop' of https://github.com/oppia/oppia into skill-opportunity-model-job

TheRealJessicaLi commented:
> @TheRealJessicaLi Hey, can you run the local devserver and run the job there and add a screenshot of a successful run in the proof part. You can generate some skills using the admin page.

@vojtechjelinek I added the screenshot. The job completes successfully,  but there is an error (included in screenshot) saying: `ERROR: "No current context. NDB calls must be made in context established by google.cloud.ndb.Client.context.": 2`, but I'm not too sure if that's relevant/what the cause of it could be. Could you take a look? Thanks!

vojtechjelinek commented on core/jobs/batch_jobs/opportunity_management_jobs.py:
```diff
@@ -43,19 +45,192 @@
+                        question_skill_links))
+            )
+            skill_opportunity.validate() # type: ignore[no-untyped-call]
+            skill_opportunity_model = opportunity_models.SkillOpportunityModel(
+                id=skill_opportunity.id,
+                skill_description=skill_opportunity.skill_description,
+                question_count=skill_opportunity.question_count
+            )
+            skill_opportunity_model.update_timestamps()
```
You will need to add `with datastore_services.get_ndb_context():`
```suggestion
            with datastore_services.get_ndb_context():
                skill_opportunity_model = opportunity_models.SkillOpportunityModel(
                    id=skill_opportunity.id,
                    skill_description=skill_opportunity.skill_description,
                    question_count=skill_opportunity.question_count
                )
                skill_opportunity_model.update_timestamps()
```

TheRealJessicaLi replied:
Done, thanks!

vojtechjelinek commented:
> > @TheRealJessicaLi Hey, can you run the local devserver and run the job there and add a screenshot of a successful run in the proof part. You can generate some skills using the admin page.
> 
> @vojtechjelinek I added the screenshot. The job completes successfully, but there is an error (included in screenshot) saying: `ERROR: "No current context. NDB calls must be made in context established by google.cloud.ndb.Client.context.": 2`, but I'm not too sure if that's relevant/what the cause of it could be. Could you take a look? Thanks!

I left a comment that should help with that error.

TheRealJessicaLi committed:
Add 'with datastore_services.get_ndb_context()'

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

TheRealJessicaLi commented:
> > > @TheRealJessicaLi Hey, can you run the local devserver and run the job there and add a screenshot of a successful run in the proof part. You can generate some skills using the admin page.
> > 
> > 
> > @vojtechjelinek I added the screenshot. The job completes successfully, but there is an error (included in screenshot) saying: `ERROR: "No current context. NDB calls must be made in context established by google.cloud.ndb.Client.context.": 2`, but I'm not too sure if that's relevant/what the cause of it could be. Could you take a look? Thanks!
> 
> I left a comment that should help with that error.

Thanks! Your suggestion fixed that error. I've updated the screenshot as well.

vojtechjelinek commented:
@TheRealJessicaLi Great, can you submit a testing request according to [the wiki](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production) and I will run the job in next few days.

TheRealJessicaLi commented:
> @TheRealJessicaLi Great, can you submit a testing request according to [the wiki](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production) and I will run the job in next few days.

Done, thanks!

vojtechjelinek reviewed (approved):
LGTM!

vojtechjelinek commented:
@TheRealJessicaLi You will need to merge from develop and also some of the tests are failing.

TheRealJessicaLi committed:
Merge branch 'develop' of https://github.com/oppia/oppia into skill-opportunity-model-job

TheRealJessicaLi committed:
linted

TheRealJessicaLi committed:
linted

TheRealJessicaLi committed:
Added test to cover failure case when validation for skill_opportunity fails

TheRealJessicaLi committed:
linted

TheRealJessicaLi committed:
Merge branch 'develop' of https://github.com/oppia/oppia into skill-opportunity-model-job

vojtechjelinek committed:
Merge branch 'develop' into skill-opportunity-model-job

vojtechjelinek committed:
Merge branch 'develop' into skill-opportunity-model-job

TheRealJessicaLi committed:
Merge branch 'develop' of https://github.com/oppia/oppia into skill-opportunity-model-job

TheRealJessicaLi committed:
Merge branch 'skill-opportunity-model-job' of https://github.com/TheRealJessicaLi/oppia into skill-opportunity-model-job
